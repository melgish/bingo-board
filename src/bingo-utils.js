import gen from "random-seed"
import shuffle from 'shuffle-array';

export const GAME = 'COVID';

// Array of indexes 0 - 4 because I'm that lazy
const BINGO = GAME.split('').map((v, k) => k);
const FREE = 2;
const RNG = gen.create();

// Array of all available bingo numbers arranged in board layout
// [1, 2, ... 15]
// [16, 17, ... 30]
// ...
export const BOARD = BINGO.map(
  i => Array.from({length: 15}, (v, k) => (i * 15) + k + 1)
);


/**
 * Create new 5 digit seed
 *
 * @return {number} random 5 digit number
 */
export function getSeed() {
  return RNG.intBetween(10000, 100000);
}

/**
 * Generate and return card for given seed
 *
 * @param {number} seed
 */
export function getCard(seed) {
  const picks = 5;
  // Use repeatable random number generator.
  const rng = gen.create(seed).random;
  // Capture 5 random values from each letter
  const cols = BINGO.map(i => shuffle.pick(BOARD[i], { picks, rng }));
  // Clear the free space
  cols[FREE][FREE] = 0;
  // Arrange in rows
  const rows = BINGO.map(i => BINGO.map(j => cols[j][i]));

  return { seed, rows };
}

