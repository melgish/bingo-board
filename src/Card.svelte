<script>
  import Ball from "./Ball.svelte";
  import Letter from "./Letter.svelte";
  import { GAME } from './bingo-utils.js';

  export let card = { rows: [], seed: 0 };
  export let calls = {};
</script>

<style>
  .card {
    display: grid;
    grid-template-columns: repeat(5, 1fr);
    grid-template-rows: repeat(5, 3rem);
    user-select: none;
    margin-bottom: 1rem;
    justify-items: stretch;
    align-content: center;
    background-color: silver;
    border: 2px solid silver;
    grid-gap: 0;
    color: black;
  }

  .card .foot {
    grid-column-start: 1;
    grid-column-end: 6;
    text-align: center;
    padding: 0.25rem 0;
  }

  .foot {
    font-weight: bold;
  }

  @media print {
    .card {
      color: black;
      border: 1px solid black;
    }
  }
</style>

<div class="card">
  {#each GAME as letter}
    <Letter {letter} />
  {/each}

  {#each card.rows as row}
    {#each row as col}
      <Ball lit={calls[col]}>{col || 'free'}</Ball>
    {/each}
  {/each}
  <div class="foot">{card.seed}</div>
</div>
