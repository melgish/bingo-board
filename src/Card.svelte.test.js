import { render } from "@testing-library/svelte"
import Card from "./Card.svelte"
import { getCard } from "./bingo-utils"

const card = getCard("12345")
const calls = {
  // free space
  0: true,
  // first box
  [card.rows[0][0]]: true,
  // last box
  [card.rows[4][4]]: true,
}

describe(Card.name, () => {
  let dom

  beforeEach(() => {
    dom = render(Card, { card, calls })
  })

  describe("when loaded", () => {
    it("should display the card", () => {
      expect(dom.getByTestId("card")).toMatchSnapshot()
      expect(dom.getAllByRole("switch").length).toBe(25)
    })
  })
})
