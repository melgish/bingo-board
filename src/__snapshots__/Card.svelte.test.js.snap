// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card when loaded should display the card 1`] = `
<div
  class="card svelte-ayymqd"
  data-testid="card"
>
  <div
    class="svelte-14jn1vn"
  >
    <span
      class="svelte-14jn1vn"
    >
      B
    </span>
  </div>
  <div
    class="svelte-14jn1vn"
  >
    <span
      class="svelte-14jn1vn"
    >
      I
    </span>
  </div>
  <div
    class="svelte-14jn1vn"
  >
    <span
      class="svelte-14jn1vn"
    >
      N
    </span>
  </div>
  <div
    class="svelte-14jn1vn"
  >
    <span
      class="svelte-14jn1vn"
    >
      G
    </span>
  </div>
  <div
    class="svelte-14jn1vn"
  >
    <span
      class="svelte-14jn1vn"
    >
      O
    </span>
  </div>
   
  <button
    aria-checked="true"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      12
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      18
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      34
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      47
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      66
    </div>
  </button>
  
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      13
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      16
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      32
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      53
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      62
    </div>
  </button>
  
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      4
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      29
    </div>
  </button>
  <button
    aria-checked="true"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      free
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      50
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      70
    </div>
  </button>
  
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      7
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      23
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      37
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      56
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      72
    </div>
  </button>
  
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      2
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      30
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      31
    </div>
  </button>
  <button
    aria-checked="false"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      49
    </div>
  </button>
  <button
    aria-checked="true"
    class="svelte-1q3vljt"
    role="switch"
  >
    <div
      class="svelte-1q3vljt"
    >
      63
    </div>
  </button>
  
   
  <div
    class="foot svelte-ayymqd"
  >
    12345
  </div>
</div>
`;
